ToolChain/Assembler/Parser/Parser.py

```diff
@@ -20,586 +19,412 @@
-        buildInstruction = Instruction(skipValidation=True)  # This is used as a model in early instruction building
-        finalCode = 0x00
-        form = None
-        possibleCodes = None
-
-        # We evaluate the line and extract any bits of relevant info to build the appropriate instruction
-        self._evaluateAndExtractInfoFromLine(line, buildInstruction, rawLine)
-
-        if type(buildInstruction.instructionCode) is not str:
-            # We assume it's an instruction (integer), since memory references or comments would be strings
-            # Parser will verify all possible codes, and generate the appropriate instruction based on its form
-            possibleCodes = self._findPossibleCodesForInstruction(buildInstruction)
-            instructionCode, instructionForm = self._getInstructionCodeAndFormUsingPossibleCodes(buildInstruction,
-                                                                                                 possibleCodes)
-            if instructionForm is None or instructionCode is None:
-                raise ValueError("Assembler general error")
-            else:
-                # Code is valid, fetch the instruction's form based on the arguments, then generate the instruction
-                buildInstruction.instructionCode = instructionCode
-                buildInstruction.instructionLength = formDescription[instructionForm]["length"]
-                lineInstruction = self._generateBinaryInstructionFromInstruction(buildInstruction,
-                                                                                 formDescription[instructionForm])
-        else:
-            # In this case, there were no instructions. Thus, we look at other possibilities
-            if COMMENT_INDICATORS in buildInstruction.instructionCode:
-                # Simply ignore a comment line
-                buildInstruction = None
-                lineInstruction = None
-
-            #  In these next cases, we can treat them as "data" variables with no real instructions.
-            #  We store the data as the instruction's source immediate value, and set instruction code to 0
-            elif DATA_ALPHA_INDICATOR in buildInstruction.instructionCode:
-                # This is a bunch of alpha data (String)
-                lineInstruction = buildInstruction.sourceImmediate
-                buildInstruction.instructionCode = 0
-
-            elif DATA_NUMERIC_INDICATOR in buildInstruction.instructionCode:
-                # This is a numeric data field (Integer)
-                lineInstruction = struct.pack(">I", buildInstruction.sourceImmediate)
-                buildInstruction.instructionCode = 0
-
-            elif DATA_MEMORY_REFERENCE in buildInstruction.instructionCode:
-                # This is a data memory reference (dataMemref)
-                lineInstruction = buildInstruction.sourceImmediate
-                buildInstruction.instructionCode = 0
+        for possibleCodes in ins:
```

Comment:
Oh! You know what I just learned something! This is indeed a nice feature that will end up simplifying the code. I guess many people are not aware of this possibility (I consider myself a decent python programmer and I just learned that now). I do think we should use for-else here as well. Information on it can be found in section 4.4 here:
https://docs.python.org/3.5/tutorial/controlflow.html

Reply:
Is this what you're thinking? It does seem to make more sense.
```
        for possibleCodes in ins:
            if insform["typeCode"] is (possibleCodes >> 4):
                self.relativeAddress += insform["length"]
                instruction += bytes((possibleCodes,))
                break
        else:
            raise ValueError("Invalid instruction format")
```

Reply:
Yup, that's what I have put in the suggestion before.